openjdk-benchmarks/src/jmh/java/org/conscrypt/benchmarks/SslEngineBenchmark.java

```diff
@@ -0,0 +1,340 @@
+import javax.net.ssl.SSLException;
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.security.NoSuchAlgorithmException;
+
+import static java.lang.Math.max;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Benchmark comparing performance of various engine implementations to conscrypt.
+ */
+@State(Scope.Benchmark)
+public class SslEngineBenchmark {
+    private static final byte[] CHARS =
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".getBytes();
+    private static final String PROTOCOL_TLS_V1_2 = "TLSv1.2";
+
+    public enum SslProvider {
+        JDK {
+            private final SSLContext clientContext = initContext(newContext(), true);
+            private final SSLContext serverContext = initContext(newContext(), false);
+
+            @Override
+            SSLEngine newClientEngine(String cipher) {
+                return initEngine(clientContext.createSSLEngine(), cipher, true);
+            }
+
+            @Override
+            SSLEngine newServerEngine(String cipher) {
+                return initEngine(serverContext.createSSLEngine(), cipher, false);
+            }
+
+            private SSLContext newContext() {
+                try {
+                    return SSLContext.getInstance(PROTOCOL_TLS_V1_2);
+                } catch (NoSuchAlgorithmException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        },
+        CONSCRYPT {
+            private final SSLContext clientContext = initContext(newContext(), true);
+            private final SSLContext serverContext = initContext(newContext(), false);
+
+            @Override
+            SSLEngine newClientEngine(String cipher) {
+                return initEngine(clientContext.createSSLEngine(), cipher, true);
+            }
+
+            @Override
+            SSLEngine newServerEngine(String cipher) {
+                return initEngine(serverContext.createSSLEngine(), cipher, false);
+            }
+
+            private SSLContext newContext() {
+                try {
+                    return SSLContext.getInstance(PROTOCOL_TLS_V1_2, new OpenSSLProvider());
+                } catch (NoSuchAlgorithmException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        },
+        NETTY {
+            private final SslContext clientContext = newClientContext();
+            private final SslContext serverContext = newServerContext();
+
+            @Override
+            SSLEngine newClientEngine(String cipher) {
+                return initEngine(
+                        clientContext.newEngine(UnpooledByteBufAllocator.DEFAULT), cipher, true);
+            }
+
+            @Override
+            SSLEngine newServerEngine(String cipher) {
+                return initEngine(
+                        serverContext.newEngine(UnpooledByteBufAllocator.DEFAULT), cipher, false);
```

Comment:
I wonder if you could also add `NETTY_POOLED` and so use the `PooledByteBufAllocator` in one benchmark. I suspect it will make a difference when heap buffers are used ;)